dnf/cli/output.py

```diff
@@ -2250,3 +2261,119 @@ def progressbar(current, total, name=None):
         sys.stdout.write('\n')
 
     sys.stdout.flush()
+
+class PackageFormatter(object):
+    """
+    # :api
+    Wrapper for dnf.package.Package, so we can control formatting
+    in --queryformat
+    """
+    def __init__(self, pkg):
+        self._pkg = pkg
+
+    def __format__(self, fmt):
+        return self.format(fmt)
+
+    def _attr(self, name):
+        name = name.lower()
+        if name not in QUERY_TAGS:
+            raise AttributeError("Unknown tag '%s'" % name)
+        return getattr(self, name, "(none)")
+
+    def __getattr__(self, attr):
+        return getattr(self._pkg, attr)
+
+    def _replace_tag(self, match):
+        subfmt = "%%%ss" % match.group(1)
+        name, fmt_fn = self._tag_format(match.group(2))
+        value = fmt_fn(self._attr(name))
+        return subfmt % (value,)
+
+    def _replace_ternary(self, match):
+        if self._attr(match.group(1)):
+            tag = match.group(2)
+        else:
+            tag = match.group(4) if match.groups() > 2 else ""
+        return tag
+
+    def _tag_format(self, tag):
+        sep = tag.find(':')
+        if sep > -1:
+            name = tag[:sep]
+            tag_format = tag[sep+1:]
+            if tag_format not in QUERY_TAG_FORMATS:
+                raise AttributeError("Unknown tag format '%s'" % tag_format)
+            format_fn = getattr(self, '_tag_format_' + tag_format, None)
+            if not format_fn:
+                raise AttributeError("Unknown tag format '%s'" % tag_format)
+        else:
+            name = tag
+            format_fn = self._tag_format_string
+        return name, format_fn
+
+    @staticmethod
+    def _tag_format_date(t):
+        return format_datetime(t)
+
+    @staticmethod
+    def _tag_format_day(t):
+        return format_date(t)
+
+    @staticmethod
+    def _tag_format_units(t):
+        return format_number(t)
+
+    @staticmethod
+    def _tag_format_string(t):
+        return ucd(t)
+
+    @staticmethod
+    def _tag_format_wrapped(t):
+        return '\n'.join(textwrap.wrap(t))
+
+    def format(self, fmt):
+        # :api
+        fmt = fmt.replace("\\n", "\n")
+        fmt = fmt.replace("\\t", "\t")
+        fmt = QUERY_TERNARY_RE.sub(self._replace_ternary, fmt)
+        return QUERY_TAG_RE.sub(self._replace_tag, fmt)
+
+    @staticmethod
+    def _reldep_to_list(obj):
```

Comment:
it's actually *from_list*. It should be something like `_reldeps_to_lines` or so.